### PR TITLE
Split XML layout to XMLELement and XmlElementBase

### LINKED ISSUE #3135: Double check name/elementname attribute for xml layout

### PR DESCRIPTION
**Pull request by 304NotModified**
Counter proposal of #3137

fixes #3135

### CODE REVIEW

File: src/NLog/Layouts/XmlElementBase.cs [hunk 1]
```diff
@@ -0,0 +1,760 @@
...
+using NLog.Internal;
+
+namespace NLog.Layouts
+{
+    /// <summary>
+    /// A specialized layout that renders XML-formatted events.
+    /// </summary>
+    [ThreadAgnostic]
+    [ThreadSafe]
+    public abstract class XmlElementBase : Layout
```
Comment (304NotModified) [inline src/NLog/Layouts/XmlElementBase.cs]: note: this is the old XmlLayout, only changes:

- elementName -> elementNameInternal and `private protected`
- samen for elementValue

**Comment by 304NotModified** [comment]
Travis:

> CSC : error CS1617: Invalid option '7.2' for /langversion; must be ISO-1, ISO-2, Default, Latest or a valid version in range 1 to 7.1. [/home/travis/build/NLog/NLog/src/NLog/NLog.csproj]

7.2 not supported?

**Comment by 304NotModified** [comment]
https://ci.appveyor.com/project/nlog/nlog/builds/22600050

> xunit.console.x86.exe :     NLog.UnitTests.Internal.AppDomainPartialTrustTests.MediumTrustWithExternalClass [FAIL]
At C:\projects\nlog\run-tests.ps1:41 char:1
+ & ${env:xunit20}\xunit.console.x86.exe .\tests\NLog.UnitTests\bin\rel ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (    NLog.UnitTe...nalClass [FAIL]:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
      File 'C:\Users\appveyor\AppData\Local\Temp\1\ff9eb82d-0972-4bcb-8f33-2ed9d35cd283\Debug.txt' doesn't exist.
      Expected: True
      Actual:   False
      Stack Trace:

**Comment by snakefoot** [comment]
You need to forward all properties from XmlLayout to XmlElement :)

**Comment by 304NotModified** [comment]
I think I forwarded all of them? The xmlayout and xmlelement has the same properties (except the element name and value)

Unit tests works locally, soms issue on appveyor (medium trust test)

**Comment by 304NotModified** [comment]
```
System.Security.SecurityException
  HResult=0x8013150A
  Message=Request for the permission of type 'System.Security.Permissions.FileIOPermission, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' failed.
  Source=mscorlib
  StackTrace:
   at System.Security.CodeAccessSecurityEngine.Check(Object demand, StackCrawlMark& stackMark, Boolean isPermSet)
   at System.Security.CodeAccessPermission.Demand()
   at System.AppDomainSetup.VerifyDir(String dir, Boolean normalize)
   at System.AppDomain.get_BaseDirectory()
   at NLog.Internal.Fakeables.AppDomainWrapper.LookupBaseDirectory(AppDomain appDomain) in D:\nlog\nlog\src\NLog\Internal\Fakeables\AppDomainWrapper.cs:line 100


System.Security.VerificationException
  HResult=0x8013150D
  Message=Operation could destabilize the runtime.
  Source=NLog
  StackTrace:
   at NLog.Internal.AppendBuilderCreator.get_Builder() in D:\nlog\nlog\src\NLog\Internal\AppendBuilderCreator.cs:line 50

```

Not sure what the issue is here, moved back to C# 7.0

**Comment by 304NotModified** [comment]
works now, not sure why MediumTrustWithExternalClass fails on C# 7.2 (AppVeyor only)

**Comment by 304NotModified** [comment]
@snakefoot please review :)

File: src/NLog/Layouts/XmlElementBase.cs [hunk 2]
```diff
@@ -0,0 +1,760 @@
...
+            Attributes = new List<XmlAttribute>();
+            Elements = new List<XmlElement>();
+            ExcludeProperties = new HashSet<string>();
+        }
+
+        /// <summary>
+        /// Name of the XML element
+        /// </summary>
+        /// <docgen category='XML Options' order='10' />
+        protected string ElementNameInternal { get => _elementName; set => _elementName = XmlHelper.XmlConvertToElementName(value?.Trim(), true); }
```
Comment (snakefoot) [inline src/NLog/Layouts/XmlElementBase.cs]: Consider making this internal instead of protected.

File: src/NLog/Layouts/XmlElementBase.cs [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+        private string _elementName;
+
+        /// <summary>
+        /// Value inside the XML element
+        /// </summary>
+        /// <docgen category='XML Options' order='10' />
+        protected Layout ElementValueInternal { get => _elementValueWrapper.Inner; set => _elementValueWrapper.Inner = value; }
```
Comment (snakefoot) [inline src/NLog/Layouts/XmlElementBase.cs]: Consider making this internal instead of protected.

**Comment by snakefoot** [comment]
Like that Elements-collection contains XmlElement-object, and like that the syntax for `<element>` now matches `<attribute>`. 

Nice improvement. Hopefully the name-clash with System.Xml.XmlElement will not hit too many.

File: src/NLog/Layouts/XmlElementBase.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/Layouts/XmlElementBase.cs]: I tried private protected ;)

**Comment by 304NotModified** [comment]
True, but that also counted for https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlattribute ;)

I will retry to enable c# 7.2 and so private protected. "internal is a bit of a hack imo"

File: src/NLog/Layouts/XmlElementBase.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/Layouts/XmlElementBase.cs]: done (with remark note added), thanks!

File: src/NLog/Layouts/XmlElementBase.cs (hunk 3 above)
Comment (304NotModified) [reply src/NLog/Layouts/XmlElementBase.cs]: done (with remark note added), thanks!

**Pull request merged**